File: grapewines_project/pipelines/03_fst/bin/plot_fst_heatmap.py
#!/usr/bin/env python3
"""
Pairwise FST heatmap.

Reads the summary table from summarize_fst.py, folds it into a square
matrix and draws it. Light cells are groups that are genetically
similar, dark cells are differentiated ones.

Input:
    pairwise_fst_summary_K<K>.tsv
    group_labels_K<K>.tsv          optional, from describe_groups.py

Output:
    pairwise_fst_heatmap_K<K>.pdf
    pairwise_fst_heatmap_K<K>.png
    pairwise_fst_matrix_K<K>.csv
"""

from __future__ import annotations

import argparse
import sys
from pathlib import Path

import matplotlib

matplotlib.use("Agg")

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from matplotlib.colors import LinearSegmentedColormap


#
# Sequential ramp built around the project's K1 blue, so the FST
# figures sit next to the ADMIXTURE barplots without clashing.
#

FST_COLOURS = [
    "#FBF7F0",
    "#BCDCE8",
    "#4E9DBC",
    "#00699A",
    "#0B3A4F",
]


def parse_args():

    parser = argparse.ArgumentParser(
        description="Heatmap of pairwise FST between ADMIXTURE groups"
    )

    parser.add_argument(
        "--summary",
        required=True,
        type=Path,
        help="pairwise_fst_summary_K<K>.tsv"
    )

    parser.add_argument(
        "--labels",
        type=Path,
        default=None,
        help="group_labels_K<K>.tsv; falls back to bare K<i>"
    )

    parser.add_argument(
        "--metric",
        default="mean_fst",
        help="Column of the summary table to plot (default mean_fst)"
    )

    parser.add_argument(
        "--k",
        default=None,
        help="K, used in the output file names "
             "(default: number of groups found)"
    )

    parser.add_argument(
        "--outdir",
        required=True,
        type=Path
    )

    return parser.parse_args()


def group_number(group: str) -> int:
    """
    K10 must sort after K9, not after K1.
    """

    return int(str(group).lstrip("K"))


def build_matrix(summary: pd.DataFrame, metric: str) -> pd.DataFrame:

    if metric not in summary.columns:
        raise ValueError(
            f"Column '{metric}' is not in the summary table. "
            f"Available: {list(summary.columns)}"
        )

    groups = sorted(
        set(summary["group_a"]) | set(summary["group_b"]),
        key=group_number
    )

    position = {group: i for i, group in enumerate(groups)}

    #
    # Filled as a plain array rather than through DataFrame.loc: recent
    # pandas hands out a read-only view from .values, so writing to the
    # diagonal in place is not portable across versions.
    #

    values = np.full((len(groups), len(groups)), np.nan)

    for row in summary.itertuples():

        i = position[row.group_a]
        j = position[row.group_b]

        value = float(getattr(row, metric))

        values[i, j] = value
        values[j, i] = value

    # A group is not differentiated from itself.
    np.fill_diagonal(values, 0.0)

    return pd.DataFrame(values, index=groups, columns=groups)


def read_labels(path: Path | None, groups: list[str]) -> list[str]:

    if path is None or not path.exists():
        return list(groups)

    lookup = pd.read_csv(path, sep="\t")

    if not {"group", "label"} <= set(lookup.columns):
        print(
            f"WARNING: {path.name} has no group/label columns; "
            f"using bare group names",
            file=sys.stderr
        )
        return list(groups)

    mapping = dict(zip(lookup["group"], lookup["label"]))

    return [mapping.get(group, group) for group in groups]


def draw(matrix: pd.DataFrame, labels: list[str], metric: str, title: str):

    n = len(matrix)

    values = matrix.to_numpy(dtype=float).copy()

    diagonal = np.eye(n, dtype=bool)

    off_diagonal = values[~diagonal]

    if np.all(np.isnan(off_diagonal)):
        raise ValueError(
            "Every off-diagonal FST value is missing; nothing to plot"
        )

    #
    # The diagonal is zero by construction and carries no information,
    # but leaving it in would stretch the colour scale from 0 and push
    # every real value into the same dark shade. Mask it, and scale the
    # ramp across the range that is actually there. Every cell is
    # labelled with its value, and the colour bar is annotated, so the
    # relative scale cannot be read as an absolute one.
    #

    values[diagonal] = np.nan

    vmin = float(np.nanmin(off_diagonal))

    vmax = float(np.nanmax(off_diagonal))

    if vmax - vmin < 1e-9:
        vmin, vmax = vmin - 1e-6, vmax + 1e-6

    colours = LinearSegmentedColormap.from_list("fst", FST_COLOURS)

    colours.set_bad("#EFEBE5")


    # Room for the labels, which carry the interpreted group names and
    # are therefore long.
    longest = max(len(label) for label in labels)

    side = 2.4 + 0.62 * n + 0.075 * longest

    figure, axes = plt.subplots(figsize=(side, side))

    image = axes.imshow(
        np.ma.masked_invalid(values),
        cmap=colours,
        vmin=vmin,
        vmax=vmax,
        interpolation="nearest"
    )


    axes.set_xticks(range(n))
    axes.set_yticks(range(n))

    axes.set_xticklabels(labels, rotation=45, ha="right", fontsize=9)
    axes.set_yticklabels(labels, fontsize=9)

    axes.set_xticks(np.arange(n + 1) - 0.5, minor=True)
    axes.set_yticks(np.arange(n + 1) - 0.5, minor=True)

    axes.grid(which="minor", color="white", linewidth=1.5)

    axes.tick_params(which="both", length=0)

    for spine in axes.spines.values():
        spine.set_visible(False)


    #
    # Print the value in every cell. The text has to flip to white on
    # the dark end of the ramp to stay readable.
    #

    for i in range(n):
        for j in range(n):

            value = values[i, j]

            if np.isnan(value):
                continue

            shade = (value - vmin) / (vmax - vmin)

            axes.text(
                j, i,
                f"{value:.4f}",
                ha="center",
                va="center",
                fontsize=8,
                color="white" if shade > 0.55 else "#222222"
            )


    axes.set_title(title, fontsize=11, pad=14)

    bar = figure.colorbar(image, ax=axes, shrink=0.72, pad=0.02)

    bar.set_label(metric, fontsize=9)

    bar.ax.tick_params(labelsize=8)

    bar.outline.set_visible(False)

    figure.tight_layout()

    return figure


def report_extremes(matrix: pd.DataFrame):

    values = matrix.to_numpy(dtype=float).copy()

    np.fill_diagonal(values, np.nan)

    if np.all(np.isnan(values)):
        return

    lowest = np.unravel_index(np.nanargmin(values), values.shape)
    highest = np.unravel_index(np.nanargmax(values), values.shape)

    groups = list(matrix.index)

    print(
        f"Lowest  FST: {groups[lowest[0]]} vs {groups[lowest[1]]} "
        f"= {values[lowest]:.4f}"
    )

    print(
        f"Highest FST: {groups[highest[0]]} vs {groups[highest[1]]} "
        f"= {values[highest]:.4f}"
    )


def main():

    args = parse_args()

    args.outdir.mkdir(parents=True, exist_ok=True)


    summary = pd.read_csv(args.summary, sep="\t")

    matrix = build_matrix(summary, args.metric)

    groups = list(matrix.index)

    k = args.k if args.k is not None else len(groups)

    labels = read_labels(args.labels, groups)


    matrix_path = args.outdir / f"pairwise_fst_matrix_K{k}.csv"

    matrix.to_csv(matrix_path, float_format="%.6f")


    title = f"Pairwise FST between ADMIXTURE groups (K = {k})"

    figure = draw(matrix, labels, args.metric, title)

    pdf_path = args.outdir / f"pairwise_fst_heatmap_K{k}.pdf"
    png_path = args.outdir / f"pairwise_fst_heatmap_K{k}.png"

    figure.savefig(pdf_path)
    figure.savefig(png_path, dpi=200)

    plt.close(figure)


    print("Wrote:")
    print(" ", matrix_path)
    print(" ", pdf_path)
    print(" ", png_path)
    print()

    report_extremes(matrix)


if __name__ == "__main__":

    try:
        main()

    except Exception as exc:
        print(f"ERROR: {exc}", file=sys.stderr)
        sys.exit(1)

File: grapewines_project/pipelines/03_fst/bin/plot_fst_heatmap.R
#!/usr/bin/env Rscript

# ---------------------------------------------------------------------
#  Pairwise FST heatmap.
#
#  Reads the summary table produced by summarize_fst.py, folds it into a
#  square matrix and draws it. Lighter cells are genetically similar
#  groups, darker cells are differentiated ones.
#
#  Written against base R only. The cluster has no guaranteed ggplot2,
#  and a missing package at the end of a 21-job FST run is an expensive
#  way to find out.
#
#  Usage:
#      plot_fst_heatmap.R --summary <tsv> --outdir <dir> [options]
# ---------------------------------------------------------------------


parse_args <- function(argv) {

    defaults <- list(
        summary = NA_character_,
        labels  = NA_character_,
        outdir  = ".",
        metric  = "mean_fst",
        k       = NA_character_
    )

    i <- 1

    while (i <= length(argv)) {

        key <- sub("^--", "", argv[i])

        if (!key %in% names(defaults)) {
            stop("Unknown argument: ", argv[i])
        }

        if (i + 1 > length(argv)) {
            stop("Missing value for ", argv[i])
        }

        defaults[[key]] <- argv[i + 1]

        i <- i + 2
    }

    if (is.na(defaults$summary)) {
        stop("--summary is required")
    }

    defaults
}


natural_group_order <- function(groups) {

    # K10 must sort after K9, not after K1.

    groups[order(as.integer(sub("^K", "", groups)))]
}


build_matrix <- function(summary, metric) {

    if (!metric %in% names(summary)) {
        stop(
            "Column '", metric, "' not in the summary table. ",
            "Available: ", paste(names(summary), collapse = ", ")
        )
    }

    groups <- natural_group_order(
        unique(c(summary$group_a, summary$group_b))
    )

    m <- matrix(
        NA_real_,
        nrow = length(groups),
        ncol = length(groups),
        dimnames = list(groups, groups)
    )

    for (row in seq_len(nrow(summary))) {

        a <- summary$group_a[row]
        b <- summary$group_b[row]
        v <- summary[[metric]][row]

        m[a, b] <- v
        m[b, a] <- v
    }

    # A group has no differentiation from itself.
    diag(m) <- 0

    m
}


# Lines of margin needed to fit the axis labels, at cex.axis = 0.8.
label_margin <- function(labels) {

    min(20, max(5, max(nchar(labels)) * 0.45))
}


draw_heatmap <- function(m, labels, metric, main_title) {

    n <- nrow(m)

    palette <- colorRampPalette(
        c("#FBF7F0", "#BCDCE8", "#4E9DBC", "#00699A", "#0B3A4F")
    )(256)

    off_diagonal <- m[upper.tri(m)]

    if (all(is.na(off_diagonal))) {
        stop("Every off-diagonal FST value is missing; nothing to plot")
    }

    limits <- c(0, max(off_diagonal, na.rm = TRUE))

    margin <- label_margin(labels)

    par(mar = c(margin, margin, 4, 7))

    image(
        x    = seq_len(n),
        y    = seq_len(n),
        z    = t(m[n:1, , drop = FALSE]),
        col  = palette,
        zlim = limits,
        axes = FALSE,
        xlab = "",
        ylab = ""
    )

    axis(
        1,
        at     = seq_len(n),
        labels = labels,
        las    = 2,
        cex.axis = 0.8,
        tick   = FALSE
    )

    axis(
        2,
        at     = seq_len(n),
        labels = rev(labels),
        las    = 2,
        cex.axis = 0.8,
        tick   = FALSE
    )

    # Cell values. White text on the dark end so it stays readable.
    for (i in seq_len(n)) {
        for (j in seq_len(n)) {

            value <- m[n - j + 1, i]

            if (is.na(value)) next

            shade <- (value - limits[1]) / diff(limits)

            text(
                i, j,
                labels = formatC(value, format = "f", digits = 4),
                cex    = 0.7,
                col    = if (shade > 0.55) "white" else "grey15"
            )
        }
    }

    box(col = "grey80")

    title(main = main_title, cex.main = 1.1)

    # Colour key in the right-hand margin.
    key_at <- seq(limits[1], limits[2], length.out = 5)

    legend(
        "topright",
        inset  = c(-0.09, 0),
        legend = formatC(rev(key_at), format = "f", digits = 3),
        fill   = rev(palette[round(seq(1, 256, length.out = 5))]),
        border = NA,
        bty    = "n",
        xpd    = TRUE,
        cex    = 0.7,
        title  = metric
    )
}


main <- function() {

    args <- parse_args(commandArgs(trailingOnly = TRUE))

    summary <- read.delim(
        args$summary,
        stringsAsFactors = FALSE
    )

    m <- build_matrix(summary, args$metric)

    groups <- rownames(m)

    k <- if (is.na(args$k)) length(groups) else args$k


    # Axis labels: interpreted names when describe_groups.py provided
    # them, bare K<i> otherwise.

    labels <- groups

    if (!is.na(args$labels) && file.exists(args$labels)) {

        lookup <- read.delim(args$labels, stringsAsFactors = FALSE)

        matched <- match(groups, lookup$group)

        labels <- ifelse(
            is.na(matched),
            groups,
            lookup$label[matched]
        )
    }


    dir.create(args$outdir, showWarnings = FALSE, recursive = TRUE)

    matrix_path <- file.path(
        args$outdir,
        sprintf("pairwise_fst_matrix_K%s.csv", k)
    )

    write.csv(
        m,
        matrix_path,
        quote = FALSE
    )


    heading <- sprintf(
        "Pairwise FST between ADMIXTURE groups (K = %s)",
        k
    )

    # Figure size = plot area + whatever the labels need. One line of
    # margin is 0.2 inch at the default point size.

    side <- 2 + 0.6 * length(groups) + 0.2 * label_margin(labels)

    pdf_path <- file.path(
        args$outdir,
        sprintf("pairwise_fst_heatmap_K%s.pdf", k)
    )

    pdf(pdf_path, width = side, height = side)
    draw_heatmap(m, labels, args$metric, heading)
    invisible(dev.off())

    png_path <- file.path(
        args$outdir,
        sprintf("pairwise_fst_heatmap_K%s.png", k)
    )

    png(png_path, width = side, height = side, units = "in", res = 200)
    draw_heatmap(m, labels, args$metric, heading)
    invisible(dev.off())


    cat("Wrote:\n")
    cat(" ", matrix_path, "\n")
    cat(" ", pdf_path, "\n")
    cat(" ", png_path, "\n")

    cat("\nLowest  FST: ")
    off <- m
    diag(off) <- NA
    lo <- which(off == min(off, na.rm = TRUE), arr.ind = TRUE)[1, ]
    cat(rownames(m)[lo[1]], "vs", colnames(m)[lo[2]],
        "=", formatC(min(off, na.rm = TRUE), format = "f", digits = 4), "\n")

    cat("Highest FST: ")
    hi <- which(off == max(off, na.rm = TRUE), arr.ind = TRUE)[1, ]
    cat(rownames(m)[hi[1]], "vs", colnames(m)[hi[2]],
        "=", formatC(max(off, na.rm = TRUE), format = "f", digits = 4), "\n")
}


main()
